### PR TITLE
CATROID-1412 Fix loudness sensor being always 0

### PR DESCRIPTION
**Pull request by wslany**
## Summary

This PR fixes the loudness sensor returning `0` by replacing its `/dev/null` recorder target with a real cache-backed recording file.

https://catrobat.atlassian.net/browse/CATROID-1412

There are several red -> green commits which are only meant to make reviewing easier. The PR should be squashed when merged.

## Red phase

The first commit adds a regression test for the loudness recorder setup and a small test seam in `SensorLoudness` so the chosen recorder path can be inspected in a unit test.

In the red state, the test fails because `SensorLoudness` still creates its recorder with `/dev/null`.

## Green phase

The second commit changes the default loudness recorder target to a file inside `Constants.SOUND_RECORDER_CACHE_DIRECTORY`.

This keeps the test green and avoids the `/dev/null` setup that appears to break `MediaRecorder.getMaxAmplitude()` on current Android/device combinations.

## Verification

- Red phase: `SensorLoudnessTest` fails before the fix
- Green phase: `SensorLoudnessTest` passes after the fix
- Manual verification on a real phone (Samsung Galaxy S24 Ultra with Android 16):
  - loudness updates correctly in the formula editor preview
  - loudness updates correctly at runtime in Catrobat programs

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Branch-and-Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed (that have passed until this PR)
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Comment by wslany** [comment]
Red phase.

Add a regression test that captures the loudness recorder target path and proves that the current `/dev/null` setup is wrong for the loudness sensor use case.

A small test seam is introduced in `SensorLoudness` so the recorder path can be observed without relying on real microphone input.

**Comment by wslany** [comment]
Green phase.

Switch the loudness sensor from `/dev/null` to a cache-backed `.m4a` file in `Constants.SOUND_RECORDER_CACHE_DIRECTORY`.

This keeps the regression test green and restores working loudness readings on a real device.

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.java [hunk 1]
```diff
@@ -101,13 +123,27 @@ public synchronized void unregisterListener(SensorCustomEventListener listener)
 						// ignored, nothing we can do
 						Log.d(TAG, "Could not stop recorder", ioException);
 					}
-					recorder = new SoundRecorder("/dev/null");
+					recorder = createSoundRecorder();
 				}
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.java]: Good point, thanks. I addressed this by deleting the loudness recorder file after the last listener unregisters and after recorder start failures, so the cache-backed workaround does not leave microphone recordings behind.

Covered in:
- `18a4edb3` test: add cleanup regression checks for loudness recorder
- `23abc0e5` fix: delete cached loudness recordings after use

File: catroid/src/test/java/org/catrobat/catroid/formulaeditor/SensorLoudnessTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.formulaeditor;
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/test/java/org/catrobat/catroid/formulaeditor/SensorLoudnessTest.java]: This was already fixed in `3d90bfad`, which moved the test to `catroid/src/test/java/org/catrobat/catroid/formulaeditor/SensorLoudnessTest.java` so the package matches the directory. The comment points to an older revision.

**Comment by harshsomankar123-tech** [comment]
@wslany The SensorLoudness fix itself looks good to me, and the added tests are helpful. 
but it also exposes an older lifecycle issue in FormulaEditorComputeDialog. The dialog sets up and starts loudness capture in setFormula(), while onStop() only unregisters the dialog itself as a SensorEventListener and never goes through the SensorHandler cleanup that stops SensorLoudness.

This likely stayed unnoticed before because SensorLoudness was using `/dev/null`, so on the affected devices the recorder path was effectively not working anyway. Now that this PR switches to a real cache-backed recording file, the missing teardown becomes visible: closing the compute dialog can leave the microphone recorder and loudness polling running until some later lifecycle cleanup happens.

Could You Please add explicit loudness/sensor cleanup when the dialog is dismissed or stopped, then this looks good to me.
Thanks!

**Review by harshsomankar123-tech — APPROVED** [review]

**Comment by harshsomankar123-tech** [comment]
@wslany LGTM!

**Comment by wslany** [comment]
Thanks @harshsomankar123-tech , it was a good catch. I added a regression test for `FormulaEditorComputeDialog.onStop()` and fixed the teardown by calling `SensorHandler.stopSensorListeners()` there, so loudness capture is now explicitly cleaned up when the dialog closes. Red-green commits to prove it.

**Comment by wslany** [comment]
Here's a screen recording with sound of an on-device test of the build: https://photos.app.goo.gl/z1ywpYWsDRW78VLa9 --- I tested: 

- working sensor in two objects in parallel
- resume from stage pause-menu
- resume when scene is continued
- restart when scene is restarted
- and that the sensor is working in the formula editor preview

**Review by dorianpercic — CHANGES_REQUESTED** [review]
Thanks for the PR! :)
I think we should implement source code files in Kotlin, since otherwise we need a refactoring ticket again.

**Comment by wslany** [comment]
> Thanks for the PR! :) I think we should implement source code files in Kotlin, since otherwise we need a refactoring ticket again.

I rewrote the touched loudness-related Java files and their dedicated tests to Kotlin in a red/green sequence:

- `2c69f34c` `test: rewrite loudness tests to Kotlin with red stubs`
- `8f4eea0a` `refactor: rewrite loudness files to Kotlin`

Verification:
- red: the rewritten targeted tests failed before the production rewrite
- green: the same targeted tests passed after the Kotlin production rewrite
- manual testing on a real phone:
  - loudness sensor works in the formula editor preview
  - loudness sensor works at runtime
  - compute dialog still works correctly
  - other compute-dialog sensor/reporter cases also still work, including time-based values like `seconds` and device sensors like inclination

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            try {
+                recorder.start()
+                statusChecker.run()
+            } catch (ioException: IOException) {
+                Log.d(TAG, "Could not start recorder", ioException)
+                listenerList.remove(listener)
+                cleanupRecorderFile()
+                recorder = createSoundRecorder()
+                return false
+            } catch (runtimeException: RuntimeException) {
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt:86]: Added a targeted @Suppress("TooGenericExceptionCaught") on registerListener() because the broad RuntimeException catch is intentional for recorder startup failures.

File: catroid/src/test/java/org/catrobat/catroid/ui/dialogs/FormulaEditorComputeDialogTest.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Test
+    fun onStopStopsSensorListeners() {
+        val bluetoothDeviceService = mock(BluetoothDeviceService::class.java)
+        val activity = Robolectric.buildActivity(AppCompatActivity::class.java).setup().get()
+        val dialog = TestFormulaEditorComputeDialog(activity, mock(Scope::class.java))
+
+        mockStatic(SensorHandler::class.java).use { sensorHandler ->
+            mockStatic(ServiceProvider::class.java).use { serviceProvider ->
+                serviceProvider.`when`<BluetoothDeviceService> {
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/test/java/org/catrobat/catroid/ui/dialogs/FormulaEditorComputeDialogTest.kt:56]: This one is already covered by the existing test setup in this module: `catroid/src/test/resources/org/powermock/extensions/configuration.properties` enables Mockito's inline mock maker (`mockito.mock-maker-class=mock-maker-inline`). The test also passes locally with `Mockito.mockStatic(...)`, so I did not change this part.

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    @VisibleForTesting
+    var statusChecker: Runnable = Runnable {
+        val loudness = (SCALE_RANGE / MAX_AMP_VALUE) * recorder.maxAmplitude
+        if (loudness != lastValue && loudness != 0.0) {
+            lastValue = loudness
+            val event = SensorCustomEvent(Sensors.LOUDNESS, loudness)
+            listenerList.forEach { it.onCustomSensorChanged(event) }
+        }
+        handler.postDelayed(statusChecker, UPDATE_INTERVAL.toLong())
+    }
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt]: Good point, thanks. I changed `statusChecker` from a mutable visible property to a private immutable `val`, so `removeCallbacks(statusChecker)` now always targets the same runnable instance.

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        private const val TAG = "SensorLoudness"
+
+        private fun defaultSoundRecorderFactory(path: String): SoundRecorder = SoundRecorder(path)
+
+        @VisibleForTesting
+        @JvmStatic
+        internal fun defaultRecorderPath(): String =
+            File(Constants.SOUND_RECORDER_CACHE_DIRECTORY, LOUDNESS_SENSOR_RECORDING_FILE_NAME).absolutePath
+
+        private const val LOUDNESS_SENSOR_RECORDING_FILE_NAME = "loudness_sensor.m4a"
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt]: Good catch. I changed the default recorder path to generate a unique per-instance file name inside `SOUND_RECORDER_CACHE_DIRECTORY`, and kept the cleanup logic so each `SensorLoudness` instance removes only its own cache file.

**Comment by wslany** [comment]
I've also tested the latest head on my phone once more: loudness sensor works everywhere, Other sensors also work in the formula editor compute dialog.

**Review by dorianpercic — APPROVED** [review]
